keps/sig-storage/4876-mutable-csinode-allocatable/README.md

```diff
@@ -0,0 +1,858 @@
+could increase API server load.
+
+- Frequent calls to a CSI driver's `NodeGetInfo` RPC
+endpoint may become expensive, particularly if the
+operation involves retrieving information from a
+remote server or performing resource-intensive tasks.
+Specifically, this is a concern at scale, where the
+cumulative cost of multiple nodes repeatedly
+querying for updates is more impactful.
+
+The risks are mitigated as follows:
+
+- The use of the Kubernetes informer pattern in the scheduler. The
+scheduler uses a `CSINode` informer and lister to
+efficiently access and watch `CSINode` objects.
+
+- Allow users to opt in to this
+feature at a per-CSI driver granularity. Administrators will
+also be able to fine-tune the interval update value as
+per their specific requirement.
+
+
+## Design Details
+
+### Feature Gate
+
+A new feature gate - `MutableCSINode` - will be introduced to control the functionality implemented by this KEP. When the
+feature gate is disabled, the `CSINode` object will remain immutable, maintaining the current behavior.
+
+### API Changes
+
+#### CSINode
+
+The `CSINode.Spec.Drivers[*].Allocatable.Count` field will be made mutable. No changes to the object structs are
+needed, only the validation logic needs to be revised. For reference, these are the API fields this KEP proposes to
+make mutable.
+
+```golang
+// CSINodeDriver holds information about the specification of one CSI driver installed
+type CSINodeDriver struct {
+    ...
+    // allocatable represents the volume resources of a node that are available for sc
+    // +optional
+    Allocatable *VolumeNodeResources
+}
+```
+
+```golang
+// VolumeNodeResources is a set of resource limits for scheduling of volumes.
+type VolumeNodeResources struct {
+    // Maximum number of unique volumes managed by the CSI driver that can be used on
+    // A volume that is both attached and mounted on a node is considered to be used o
+    // The same rule applies for a unique volume that is shared among multiple pods on
+    // If this field is not specified, then the supported number of volumes on this no
+    // +optional
+    Count *int32
+}
+```
+
+#### CSIDriver
+
+A new field, `AllocatableUpdateInterval`, will be added to the `CSIDriverSpec` struct. This field allows a CSI driver to
+specify the interval at which the Kubelet should periodically query a driver's `NodeGetInfo` RPC endpoint to update the
+`CSINode` object. If this field is not set, updates will only occur in response to volume attachment failures as a result of no
+capacity.
+
+```golang
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+    ...
+    // AllocatableUpdateInterval specifies the interval between periodic updates of
+    // the CSINode allocatable capacity for this driver. If not set, periodic updates
+    // are disabled, and updates occur only upon detecting capacity-related failures.
+    // +optional
+    AllocatableUpdateInterval *metav1.Duration
```

Comment:
Would it be a fixed interval period only? I think it'd be useful to do this with exp backoff too.

Reply:
Yes, a fixed interval period only. I'm not sure that it would make sense to update this value using an exponential backoff strategy, but open to the idea.